modules/core/src/main/java/org/apache/synapse/aspects/flow/statistics/opentracing/management/parentresolving/ArtifactHolderBasedParentResolver.java

```diff
@@ -0,0 +1,271 @@
+     * Returns null, when none is obtained.
+     * @param structuringElementStack   Structuring element stack of an artifact holder.
+     * @return                          Parent component unique id.
+     */
+    private static String getReportedParentComponentUniqueId(Stack<StructuringElement> structuringElementStack) {
+        String parentComponentUniqueId = null;
+        if (structuringElementStack != null && !structuringElementStack.isEmpty()) {
+            parentComponentUniqueId = structuringElementStack.peek().getId();
+        }
+        return parentComponentUniqueId;
+    }
+
+    /**
+     * Returns the span wrapper which matches to the given component unique id.
+     * When the span wrapper denoted by the component unique id contains anonymous sequences, suitable one among those
+     * will be chosen and returned.
+     * Otherwise, it will be examined whether is there any other copy exists with the same parent component unique id,
+     * and the reference to the suitable one will be returned.
+     * @param parentComponentUniqueId
+     * @param child
+     * @param spanStore
+     * @param childSynCtx
+     * @return
+     */
+    private static SpanWrapper findSpanWrapperByComponentUniqueId(String parentComponentUniqueId,
+                                                                  StatisticDataUnit child,
+                                                                  SpanStore spanStore,
+                                                                  MessageContext childSynCtx) {
+        SpanWrapper parent = spanStore.getSpanWrapperByComponentUniqueId(parentComponentUniqueId);
+        if (parent != null) {
+            if (!parent.getAnonymousSequences().isEmpty()) {
+                /*
+                The chosen parent contains anonymous sequences, which means that
+                the actual parent should be an appropriate anonymous sequence.
+                */
+                return resolveAnonymousSequence(parent, child, spanStore, childSynCtx);
+            }
+
+            /*
+            In cases like the the following one - where multiple copies of a proxy out sequence are created,
+            the parent won't just be the one identified with the component unique id,
+            but one of its copies with the same parent unique id.
+
+            <target>
+                <inSequence> <iterate> ... </iterate> </inSequence>
+                <outSequence> <aggregate> ... </aggregate> </outSequence>
+            </target>
+
+            In such cases,
+            find the appropriate copy of the parent
+            which doesn't already contain an element with the same child unique id as a child.
+            */
+            if (!isAlreadyAParent(parent, child.getComponentId())) {
+                return parent;
+            }
+            for (SpanWrapper spanWrapper : spanStore.getSpanWrappers().values()) {
```

Comment:
can we have a NPE here?

Reply:
No.

`spanStore` is passed down from [this](https://github.com/wso2/wso2-synapse/pull/1466/files#diff-2a3b98173b41bb95122e846bf31228b2R101) point.

When we are initializing a `TracingScope` object [here](https://github.com/wso2/wso2-synapse/pull/1466/files#diff-2a3b98173b41bb95122e846bf31228b2R98), a new `SpanStore` object is created [like this](https://github.com/wso2/wso2-synapse/pull/1466/files#diff-6fec681eedf33528947be8853b7feaf5R62).